src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetService.cs

```diff
@@ -60,6 +62,28 @@ public class DataSetService(
             });
     }
 
+    public async Task<Either<ActionResult, DataSetSummaryViewModel>> GetDataSet(
+        Guid dataSetId,
+        CancellationToken cancellationToken = default)
+    {
+        return await CheckDataSetExists(dataSetId, cancellationToken)
+            .OnSuccessCombineWith(dataSet => CheckPublicationExists(dataSet.PublicationId, cancellationToken))
+            .OnSuccess(async combinedDataSetAndPublication =>
+            {
+                (var dataSet, var publication) = combinedDataSetAndPublication;
+
+                return await userService.CheckCanViewPublication(publication)
+                    .OnSuccess(_ => dataSet);
+            })
+            .OnSuccessCombineWith(async dataSet => await GetLatestDataSetVersionFilesByFileId(dataSet, cancellationToken))
+            .OnSuccess(combinedDataSetAndReleaseFiles =>
+            {
+                (var dataSet, var releaseFilesByFileId) = combinedDataSetAndReleaseFiles;
+
+                return MapDataSet(dataSet, releaseFilesByFileId);
+            });
```

Comment:
You could do this which might be slightly simpler because it only involves one `OnSuccessCombineWith`

```csharp
        return await CheckDataSetExists(dataSetId, cancellationToken)
            .OnSuccessDo(async dataSet => await CheckPublicationExists(dataSet.PublicationId, cancellationToken)
                .OnSuccess(userService.CheckCanViewPublication))
            .OnSuccessCombineWith(async dataSet => await GetLatestDataSetVersionFilesByFileId(dataSet, cancellationToken))
            .OnSuccess(combinedDataSetAndReleaseFiles =>
            {
                var (dataSet, releaseFilesByFileId) = combinedDataSetAndReleaseFiles;
                return MapDataSet(dataSet, releaseFilesByFileId);
            });
```

Reply:
Nice! Didn't think of that